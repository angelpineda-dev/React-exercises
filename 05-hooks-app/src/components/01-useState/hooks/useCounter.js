import { useState } from "react";

/* Custom hook no son mÃ¡s que simples funciones */
export const useCounter = (initialState = 1) => {
  const [counter, setCounter] = useState(initialState);

  const increment = () => {
    setCounter(counter + 1);
  };

  const decrement = () => {
    setCounter(counter - 1);
  };

  const reset = () => {
    setCounter(initialState);
  };

  return {
    counter,
    increment,
    decrement,
    reset,
  };
};
