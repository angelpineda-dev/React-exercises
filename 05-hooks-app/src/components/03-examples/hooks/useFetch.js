import { useState, useEffect, useRef } from "react";

export const useFetch = (url) => {
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  const isMounted = useRef(true);

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState({
      data: null,
      loading: true,
      error: null,
    });

    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        setTimeout(() => {
          if (isMounted) {
            setState({
              loading: false,
              error: null,
              data,
            });
          }
        }, 1);
      });
  }, [url]);

  return state;
};
