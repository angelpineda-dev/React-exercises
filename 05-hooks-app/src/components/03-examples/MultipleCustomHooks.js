import { useCounter } from "../01-useState/hooks/useCounter";
import { useFetch } from "./hooks/useFetch";

const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1);

  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  /* 
    null retorna null
    !null retorna true
    !!null retorna false
  */

  let { author, quote } = !!data && data[0];

  return (
    <div>
      <h2>Multiple Custom Hooks!</h2>
      <br />

      {loading ? (
        <div className="alert alert-info text-center">Loading...</div>
      ) : (
        <blockquote className="blockquote text-end">
          <p className="mb-3">{quote}</p>
          <footer className="blockquote-footer">{author}</footer>
        </blockquote>
      )}

      <button className="btn btn-primary" onClick={increment}>
        Siguiente frase
      </button>
    </div>
  );
};

export default MultipleCustomHooks;
