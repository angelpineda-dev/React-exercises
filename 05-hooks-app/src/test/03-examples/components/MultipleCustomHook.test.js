import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks";
import { shallow } from "enzyme";
import { useFetch } from "../../../components/03-examples/hooks/useFetch";
import { useCounter } from "../../../components/01-useState/hooks/useCounter";

jest.mock("../../../components/03-examples/hooks/useFetch");
jest.mock("../../../components/01-useState/hooks/useCounter");

describe("Pruebas en <MultipleCustomHook/>", () => {
  beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 10,
      increment: () => {},
    });
  });

  test("Debe mostar el componente correctamente", () => {
    useFetch.mockReturnValue({
      data: null,
      loagin: true,
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar la informaciÃ³n", () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: "Angel",
          quote: "Hola Mundo",
        },
      ],
      loading: false,
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper.find(".alert").exists()).toBe(false);
    expect(wrapper.find(".mb-3").text().trim()).toBe("Hola Mundo");
    expect(wrapper.find("footer").text().trim()).toBe("Angel");
  });
});
