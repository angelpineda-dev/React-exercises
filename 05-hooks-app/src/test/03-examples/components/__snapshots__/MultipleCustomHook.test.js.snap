// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHook/> Debe mostar el componente correctamente 1`] = `
<div>
  <h2>
    Multiple Custom Hooks!
  </h2>
  <br />
  <blockquote
    className="blockquote text-end"
  >
    <p
      className="mb-3"
    />
    <footer
      className="blockquote-footer"
    />
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</div>
`;
